[BUG] Build fails when BPFTIME_BUILD_WITH_LIBBPF=OFF

## Describe the Bug

Building bpftime with `-DBPFTIME_BUILD_WITH_LIBBPF=OFF` fails with multiple compilation errors.
This mode is documented in `installation.md` (lines 114–117, 169–176) and has a dedicated
`build-wo-libbpf` target in the `Makefile`, but is completely untested in CI
— all 13 CI workflow files build exclusively with `BPFTIME_BUILD_WITH_LIBBPF=ON` (the default).

As a result, about 7 source/CMake files have accumulated unconditional libbpf dependencies
that break the no-libbpf build path. This affects:

- macOS users (no kernel eBPF)
- Pure userspace deployments
- Porting efforts to non-Linux platforms

## Steps to Reproduce

```bash
# System: Ubuntu 22.04, GCC 12, Clang 17, LLVM 17
cmake -B build \
  -DCMAKE_BUILD_TYPE=Release \
  -DBPFTIME_LLVM_JIT=1 \
  -DBPFTIME_BUILD_WITH_LIBBPF=OFF
cmake --build build -j$(nproc)
```

## Actual Behavior

Seven distinct build failures, in order of occurrence:

### 1. `tools/aot/CMakeLists.txt` — unconditional `libbpf` dependency

```
CMake Error at tools/aot/CMakeLists.txt:41 (add_dependencies):
  The dependency target "libbpf" of target "bpftime-aot-cli" does not exist.
```

`add_dependencies(bpftime-aot-cli ... libbpf)` is called unconditionally,
but the `libbpf` target is never created when `BPFTIME_BUILD_WITH_LIBBPF=OFF`.

### 2. `tools/aot/CMakeLists.txt` — hardcoded LLVM library path

```cmake
target_link_libraries(bpftime-aot-cli PRIVATE -L/usr/lib/llvm-20/lib/ ...)
```

`/usr/lib/llvm-20/lib/` does not exist on systems with LLVM 17 or other versions.
The LLVM library directory is already available via `${LLVM_LIBRARY_DIRS}`.

### 3. `queue.hpp`, `stack.hpp`, `bloom_filter.hpp` — unconditional `#include <bpf/bpf.h>`

```
fatal error: bpf/bpf.h: No such file or directory
  #include <bpf/bpf.h>
```

These three header files include `<bpf/bpf.h>` solely for three trivial integer constants:

```c
#define BPF_ANY     0
#define BPF_NOEXIST 1
#define BPF_EXIST   2
```

Other parts of the project already avoid this dependency by using inline comments
(`/*BPF_ANY*/`, `/*BPF_EXIST*/` in `map_common_def.hpp:88–89` and
`lpm_trie_map.cpp:279–280`). The queue/stack/bloom_filter map types were
added later and did not follow this convention.

### 4. `prog_array.cpp` — unconditional libbpf headers on Linux

```
fatal error: bpf/bpf.h: No such file or directory
```

`prog_array.cpp:8–10` unconditionally includes `<bpf/bpf.h>`, `<linux/bpf.h>`,
and `<bpf/libbpf.h>` under `#if __linux__`. The kernel-interaction functions
(`my_bpf_obj_get_info_by_fd`, `my_bpf_prog_get_fd_by_id`, and the kernel-fd
branches of `elem_lookup`/`elem_update`) should be guarded by
`#ifdef BPFTIME_BUILD_WITH_LIBBPF`.

### 5. `syscall_context.cpp` — unconditional `#include <bpf/bpf.h>` + missing `bpf_obj_get_info_by_fd`

```
error: 'bpf_obj_get_info_by_fd' was not declared in this scope
```

When `<bpf/bpf.h>` is not available (guarded by `#ifdef BPFTIME_BUILD_WITH_LIBBPF`),
the convenience function `bpf_obj_get_info_by_fd` is not declared. A local
fallback using direct `syscall(__NR_bpf, BPF_OBJ_GET_INFO_BY_FD, ...)` is needed.

### 6. `syscall_context.cpp` — `map_extra` field not in system kernel headers

```
error: 'union bpf_attr' has no member named 'map_extra'
error: 'struct bpf_map_info' has no member named 'map_extra'
```

The system `/usr/include/linux/bpf.h` (from kernel 5.15) does not have the
`map_extra` field added in newer kernels. The libbpf submodule's uapi copy
(`third_party/libbpf/include/uapi/linux/bpf.h`) has it, but is not included
in the include path when `BPFTIME_BUILD_WITH_LIBBPF=OFF`.

### 7. `tools/CMakeLists.txt` — AOT CLI built without libbpf

The AOT CLI depends heavily on libbpf (`bpf_object__open`, `bpf_program__name`,
`libbpf_set_print`) and should be excluded when `BPFTIME_BUILD_WITH_LIBBPF=OFF`.

## Expected Behavior

Build completes successfully with `BPFTIME_BUILD_WITH_LIBBPF=OFF`, producing
a fully functional userspace-only bpftime runtime (VM, maps, attach mechanisms,
CLI tool, syscall server) without any kernel eBPF dependency.

## Environment

| Item | Value |
|------|-------|
| OS | Ubuntu 22.04.5 LTS (WSL2) |
| Kernel | 5.15.167.4-microsoft-standard-WSL2 |
| Architecture | x86_64 |
| GCC | 12.4.0 |
| Clang | 17.0.6 |
| LLVM | 17.0.6 |
| CMake | 3.22.1 |
| Boost | 1.74.0 |

## Root Cause Analysis

The `BPFTIME_BUILD_WITH_LIBBPF=OFF` build mode:

1. **Is documented** — `installation.md` lines 114–117, 169–176 describe it as the
   mode for macOS and portability
2. **Has conditional code** — 25+ source files already have `#ifdef BPFTIME_BUILD_WITH_LIBBPF`
   guards, showing the architecture was designed for this
3. **Has zero CI coverage** — all 13 CI workflows build only with `ON` (default)

This is the standard pattern of "bit rot in untested code paths." Every new feature
(queue/stack/bloom filters, AOT CLI, `map_extra` support) was tested only with the
default `ON` path, and no CI job catches regressions in the `OFF` path.

## Proposed Fix

All fixes share the single root cause (untested no-libbpf path) and can go in one PR.
Summary of changes required:

| File | Fix |
|------|-----|
| `runtime/src/bpf_map/userspace/queue.hpp` | Replace `#include <bpf/bpf.h>` with local `#define BPF_ANY/BPF_NOEXIST/BPF_EXIST` |
| `runtime/src/bpf_map/userspace/stack.hpp` | Same |
| `runtime/src/bpf_map/userspace/bloom_filter.hpp` | Same |
| `runtime/src/bpf_map/userspace/prog_array.cpp` | Guard kernel-interaction code with `#ifdef BPFTIME_BUILD_WITH_LIBBPF` |
| `runtime/syscall-server/syscall_context.cpp` | Guard `<bpf/bpf.h>` include + add fallback `bpf_obj_get_info_by_fd` using direct syscall |
| `tools/aot/CMakeLists.txt` | Guard `libbpf` dependency + replace hardcoded `-L/usr/lib/llvm-20/lib/` with `target_link_directories(... ${LLVM_LIBRARY_DIRS})` |
| `tools/CMakeLists.txt` | Exclude AOT subdirectory when `BPFTIME_BUILD_WITH_LIBBPF=OFF` |
| `runtime/syscall-server/CMakeLists.txt` | Always add libbpf submodule uapi include path for newer kernel header definitions |

Build verified: all targets compile successfully with `-DBPFTIME_BUILD_WITH_LIBBPF=OFF -DBPFTIME_LLVM_JIT=1`.

## Additional Recommendation

Add a CI job that builds with `-DBPFTIME_BUILD_WITH_LIBBPF=OFF` as a compile-only
smoke test (e.g., extend `build-gcc13.yml` or add a new workflow). This would
prevent future regressions in the no-libbpf path and serve as a canary for
macOS/portability efforts.

--- a/.github/workflows/build-gcc13.yml
+++ b/.github/workflows/build-gcc13.yml
@@ -87,8 +87,10 @@ jobs:
             -DBPFTIME_BUILD_WITH_LIBBPF=OFF \
             -DBPFTIME_ENABLE_UNIT_TESTING=OFF
           cmake --build build-no-libbpf --target install -j$(nproc)
-          readelf -Ws build-no-libbpf/runtime/agent/libbpftime-agent.so | grep -q register_llvm_vm_factory
-          readelf -Ws build-no-libbpf/runtime/agent/libbpftime-agent.so | grep -q create_ubpf_vm_instance
+          readelf -Ws build-no-libbpf/runtime/agent/libbpftime-agent.so | grep register_llvm_vm_factory
+          readelf -Ws build-no-libbpf/runtime/agent/libbpftime-agent.so | grep create_ubpf_vm_instance
+          readelf -Ws build-no-libbpf/runtime/syscall-server/libbpftime-syscall-server.so | grep register_llvm_vm_factory
+          readelf -Ws build-no-libbpf/runtime/syscall-server/libbpftime-syscall-server.so | grep create_ubpf_vm_instance
 
       - name: Configure daemon build and tests
         env:

--- a/.github/workflows/build-gcc13.yml
+++ b/.github/workflows/build-gcc13.yml
@@ -77,6 +77,20 @@ jobs:
       - name: Build
         run: cmake --build build -j$(nproc)
 
+      - name: Build without libbpf
+        env:
+          CC: gcc-13
+          CXX: g++-13
+        run: |
+          cmake -S . -B build-no-libbpf \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DCMAKE_INSTALL_PREFIX=$PWD/build-no-libbpf/install \
+            -DBPFTIME_BUILD_WITH_LIBBPF=OFF \
+            -DBPFTIME_ENABLE_UNIT_TESTING=OFF
+          cmake --build build-no-libbpf --target install -j$(nproc)
+          readelf -Ws build-no-libbpf/runtime/agent/libbpftime-agent.so | grep -q register_llvm_vm_factory
+          readelf -Ws build-no-libbpf/runtime/agent/libbpftime-agent.so | grep -q create_ubpf_vm_instance
+
       - name: Configure daemon build and tests
         env:
           CC: gcc-13

--- a/.github/workflows/build-gcc13.yml
+++ b/.github/workflows/build-gcc13.yml
@@ -84,7 +84,6 @@ jobs:
         run: |
           cmake -S . -B build-no-libbpf \
             -DCMAKE_BUILD_TYPE=Release \
-            -DCMAKE_INSTALL_PREFIX=$PWD/build-no-libbpf/install \
             -DBPFTIME_BUILD_WITH_LIBBPF=OFF \
             -DBPFTIME_ENABLE_UNIT_TESTING=OFF
           cmake --build build-no-libbpf --target install -j$(nproc)

--- a/runtime/agent/CMakeLists.txt
+++ b/runtime/agent/CMakeLists.txt
@@ -29,7 +29,6 @@ endif()
 
 if(${BPFTIME_LLVM_JIT})
   add_dependencies(bpftime-agent bpftime_llvm_vm llvmbpf_vm)
-  target_link_options(bpftime-agent PUBLIC "-Wl,--whole-archive" "$<TARGET_FILE:bpftime_llvm_vm>" "-Wl,--no-whole-archive")
   target_link_libraries(bpftime-agent PUBLIC bpftime_llvm_vm)
   # Statically embed the compat_llvm constructor TU to guarantee VM factory registration in LD_PRELOAD context
   target_sources(bpftime-agent PRIVATE $<TARGET_OBJECTS:bpftime_llvm_vm_obj>)

--- a/runtime/src/bpf_map/userspace/bloom_filter.cpp
+++ b/runtime/src/bpf_map/userspace/bloom_filter.cpp
@@ -1,4 +1,5 @@
 #include <bpf_map/userspace/bloom_filter.hpp>
+#include "linux/bpf.h"
 #include <spdlog/spdlog.h>
 #include <cerrno>
 #include <cstring>

--- a/runtime/src/bpf_map/userspace/bloom_filter.hpp
+++ b/runtime/src/bpf_map/userspace/bloom_filter.hpp
@@ -24,7 +24,6 @@
 #include <boost/interprocess/sync/interprocess_mutex.hpp>
 #include <boost/interprocess/containers/vector.hpp>
 #include <boost/interprocess/allocators/allocator.hpp>
-#include <bpf/bpf.h>
 #include <cstdint>
 
 namespace bpftime

--- a/runtime/src/bpf_map/userspace/prog_array.cpp
+++ b/runtime/src/bpf_map/userspace/prog_array.cpp
@@ -5,9 +5,7 @@
  */
 
 #if __linux__
-#include "bpf/bpf.h"
 #include "linux/bpf.h"
-#include <bpf/libbpf.h>
 #include <gnu/lib-names.h>
 #include <asm/unistd.h>
 #elif __APPLE__

--- a/runtime/src/bpf_map/userspace/queue.cpp
+++ b/runtime/src/bpf_map/userspace/queue.cpp
@@ -1,4 +1,5 @@
 #include <bpf_map/userspace/queue.hpp>
+#include "linux/bpf.h"
 #include <spdlog/spdlog.h>
 #include <cerrno>
 #include <cstring>

--- a/runtime/src/bpf_map/userspace/queue.hpp
+++ b/runtime/src/bpf_map/userspace/queue.hpp
@@ -20,8 +20,6 @@
 #include <boost/interprocess/managed_shared_memory.hpp>
 #include <boost/interprocess/containers/vector.hpp>
 #include <boost/interprocess/allocators/allocator.hpp>
-// Include bpf.h for BPF flags like BPF_ANY, BPF_EXIST, etc.
-#include <bpf/bpf.h>
 
 // Forward declaration
 namespace bpftime

--- a/runtime/src/bpf_map/userspace/stack.cpp
+++ b/runtime/src/bpf_map/userspace/stack.cpp
@@ -1,4 +1,5 @@
 #include <bpf_map/userspace/stack.hpp>
+#include "linux/bpf.h"
 #include <spdlog/spdlog.h>
 #include <cerrno>
 #include <cstring>

--- a/runtime/src/bpf_map/userspace/stack.hpp
+++ b/runtime/src/bpf_map/userspace/stack.hpp
@@ -25,8 +25,6 @@
 #include <boost/interprocess/sync/interprocess_mutex.hpp>
 #include <boost/interprocess/containers/vector.hpp>
 #include <boost/interprocess/allocators/allocator.hpp>
-// Include bpf.h for BPF flags like BPF_ANY, BPF_EXIST, etc.
-#include <bpf/bpf.h>
 
 // Forward declaration
 namespace bpftime

--- a/runtime/syscall-server/CMakeLists.txt
+++ b/runtime/syscall-server/CMakeLists.txt
@@ -33,21 +33,17 @@ target_link_libraries(bpftime-syscall-server PUBLIC
 
 add_dependencies(bpftime-syscall-server spdlog::spdlog)
 
-if(${BPFTIME_BUILD_WITH_LIBBPF})
-
 target_include_directories(bpftime-syscall-server
     PUBLIC
     "../include"
     "../../vm"
-    "../../third_party/libbpf/include"
     "../../third_party/libbpf/include/uapi"
     ${SPDLOG_INCLUDE}
 )
-else()
-target_include_directories(bpftime-syscall-server
-    PUBLIC
-    "../../vm"
-    ${SPDLOG_INCLUDE}
+
+if(${BPFTIME_BUILD_WITH_LIBBPF})
+target_include_directories(bpftime-syscall-server PUBLIC
+    "../../third_party/libbpf/include"
 )
 endif()
 

--- a/runtime/syscall-server/CMakeLists.txt
+++ b/runtime/syscall-server/CMakeLists.txt
@@ -20,8 +20,8 @@ endif()
 
 if(${BPFTIME_LLVM_JIT})
   add_dependencies(bpftime-syscall-server bpftime_llvm_vm llvmbpf_vm)
-  target_link_options(bpftime-syscall-server PUBLIC "-Wl,--whole-archive" "$<TARGET_FILE:bpftime_llvm_vm>" "-Wl,--no-whole-archive")
   target_link_libraries(bpftime-syscall-server PUBLIC bpftime_llvm_vm)
+  target_sources(bpftime-syscall-server PRIVATE $<TARGET_OBJECTS:bpftime_llvm_vm_obj>)
 endif()
 
 target_link_libraries(bpftime-syscall-server PUBLIC

--- a/runtime/syscall-server/syscall_context.cpp
+++ b/runtime/syscall-server/syscall_context.cpp
@@ -72,7 +72,7 @@ namespace {
 [[noreturn]] void
 exit_for_startup_allocation_failure(const std::exception &error)
 {
-	auto config = bpftime::construct_agent_config_from_env();
+	auto config = bpftime::construct_runtime_config_from_env();
 	SPDLOG_CRITICAL(
 		"Unable to initialize bpftime shared memory ({} MiB, {} fd slots): {}",
 		config.shm_memory_size, config.max_fd_count, error.what());

--- a/runtime/syscall-server/syscall_context.cpp
+++ b/runtime/syscall-server/syscall_context.cpp
@@ -24,6 +24,7 @@
 #if __linux__
 #include "linux/perf_event.h"
 #include <linux/bpf.h>
+#include <asm/unistd.h>
 #include <sys/epoll.h>
 #ifdef BPFTIME_BUILD_WITH_LIBBPF
 #include <bpf/bpf.h>
@@ -300,7 +301,7 @@ int syscall_context::create_kernel_bpf_map(int map_fd, int bpftime_map_type)
 	bpf_map_info info = {};
 	uint32_t info_len = sizeof(info);
 	int res = get_bpf_obj_info_by_fd(map_fd, &info, &info_len,
-				 orig_syscall_fn);
+					 orig_syscall_fn);
 	if (res < 0) {
 		SPDLOG_ERROR("Failed to get map info for fd {}", map_fd);
 		return -1;
@@ -381,9 +382,10 @@ int syscall_context::create_kernel_bpf_prog_in_userspace(int cmd,
 			if (inst.src_reg == 1 || inst.src_reg == 2) {
 				bpf_map_info info = {};
 				uint32_t info_len = sizeof(info);
-				int res = get_bpf_obj_info_by_fd(
-					inst.imm, &info, &info_len,
-					orig_syscall_fn);
+				int res =
+					get_bpf_obj_info_by_fd(inst.imm, &info,
+							       &info_len,
+							       orig_syscall_fn);
 				if (res < 0) {
 					SPDLOG_ERROR(
 						"Failed to get map info for id {}",

--- a/runtime/syscall-server/syscall_context.cpp
+++ b/runtime/syscall-server/syscall_context.cpp
@@ -25,7 +25,9 @@
 #include "linux/perf_event.h"
 #include <linux/bpf.h>
 #include <sys/epoll.h>
+#ifdef BPFTIME_BUILD_WITH_LIBBPF
 #include <bpf/bpf.h>
+#endif
 #include <linux/perf_event.h>
 #include <linux/filter.h>
 #elif __APPLE__
@@ -72,13 +74,34 @@ namespace {
 [[noreturn]] void
 exit_for_startup_allocation_failure(const std::exception &error)
 {
-	auto config = bpftime::construct_agent_config_from_env();
+	auto config = bpftime::construct_runtime_config_from_env();
 	SPDLOG_CRITICAL(
 		"Unable to initialize bpftime shared memory ({} MiB, {} fd slots): {}",
 		config.shm_memory_size, config.max_fd_count, error.what());
 	SPDLOG_CRITICAL(
 		"Increase BPFTIME_SHM_MEMORY_MB or decrease BPFTIME_MAX_FD_COUNT");
 	std::exit(EXIT_FAILURE);
+}
+
+int get_bpf_obj_info_by_fd(int fd, void *info, uint32_t *info_len,
+			   long (*syscall_fn)(long, ...))
+{
+#ifdef BPFTIME_BUILD_WITH_LIBBPF
+	return bpf_obj_get_info_by_fd(fd, info, info_len);
+#elif defined(__linux__)
+	union bpf_attr attr = {};
+	attr.info.bpf_fd = fd;
+	attr.info.info_len = *info_len;
+	attr.info.info = reinterpret_cast<uintptr_t>(info);
+	int result = syscall_fn(__NR_bpf, BPF_OBJ_GET_INFO_BY_FD, &attr,
+				sizeof(attr.info));
+	if (result == 0)
+		*info_len = attr.info.info_len;
+	return result;
+#else
+	errno = ENOTSUP;
+	return -1;
+#endif
 }
 
 std::string format_verifier_program_dump(const uint64_t *instructions,
@@ -276,7 +299,8 @@ int syscall_context::create_kernel_bpf_map(int map_fd, int bpftime_map_type)
 {
 	bpf_map_info info = {};
 	uint32_t info_len = sizeof(info);
-	int res = bpf_obj_get_info_by_fd(map_fd, &info, &info_len);
+	int res = get_bpf_obj_info_by_fd(map_fd, &info, &info_len,
+				 orig_syscall_fn);
 	if (res < 0) {
 		SPDLOG_ERROR("Failed to get map info for fd {}", map_fd);
 		return -1;
@@ -357,8 +381,9 @@ int syscall_context::create_kernel_bpf_prog_in_userspace(int cmd,
 			if (inst.src_reg == 1 || inst.src_reg == 2) {
 				bpf_map_info info = {};
 				uint32_t info_len = sizeof(info);
-				int res = bpf_obj_get_info_by_fd(
-					inst.imm, &info, &info_len);
+				int res = get_bpf_obj_info_by_fd(
+					inst.imm, &info, &info_len,
+					orig_syscall_fn);
 				if (res < 0) {
 					SPDLOG_ERROR(
 						"Failed to get map info for id {}",

--- a/runtime/unit-test/maps/test_bloom_filter_map.cpp
+++ b/runtime/unit-test/maps/test_bloom_filter_map.cpp
@@ -3,6 +3,7 @@
 
 #include "bpf_map/userspace/bloom_filter.hpp"
 #include "../common_def.hpp"
+#include <linux/bpf.h>
 #include <boost/interprocess/managed_shared_memory.hpp>
 #include <vector>
 #include <cstring>

--- a/runtime/unit-test/maps/test_queue_map.cpp
+++ b/runtime/unit-test/maps/test_queue_map.cpp
@@ -5,6 +5,7 @@
 
 #include "bpf_map/userspace/queue.hpp" // Adjust path as needed
 #include "../common_def.hpp"
+#include <linux/bpf.h>
 #include <boost/interprocess/managed_shared_memory.hpp>
 #include <boost/interprocess/containers/string.hpp>
 #include <vector>

--- a/runtime/unit-test/maps/test_stack_map.cpp
+++ b/runtime/unit-test/maps/test_stack_map.cpp
@@ -3,6 +3,7 @@
 
 #include "bpf_map/userspace/stack.hpp" // Adjust path as needed
 #include "../common_def.hpp"
+#include <linux/bpf.h>
 #include <boost/interprocess/managed_shared_memory.hpp>
 #include <boost/interprocess/containers/string.hpp>
 #include <vector>

--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -1,6 +1,6 @@
 add_subdirectory(bpftimetool)
 add_subdirectory(cli)
-if(BPFTIME_LLVM_JIT)
+if(BPFTIME_LLVM_JIT AND BPFTIME_BUILD_WITH_LIBBPF)
   message(STATUS "Using llvm-jit")
   add_subdirectory(aot)
 endif()

--- a/vm/compat/llvm-vm/CMakeLists.txt
+++ b/vm/compat/llvm-vm/CMakeLists.txt
@@ -7,5 +7,8 @@ target_link_libraries(bpftime_llvm_vm PUBLIC bpftime_vm_compat spdlog::spdlog ll
 
 target_include_directories(bpftime_llvm_vm_obj PRIVATE ../../llvm-jit/include ${CMAKE_CURRENT_SOURCE_DIR} ${SPDLOG_INCLUDE} ../include)
 
-add_subdirectory(../../llvm-jit ${CMAKE_CURRENT_BINARY_DIR}/llvm-jit)
-
+if(BPFTIME_BUILD_WITH_LIBBPF)
+  add_subdirectory(../../llvm-jit ${CMAKE_CURRENT_BINARY_DIR}/llvm-jit)
+else()
+  add_subdirectory(../../llvm-jit ${CMAKE_CURRENT_BINARY_DIR}/llvm-jit EXCLUDE_FROM_ALL)
+endif()

